### PR TITLE
Add ram pkg

### PR DESCRIPTION
**Pull request by heww**
Add RAM pkg to implement HasPermission in RBAC mode

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/common/ram/cabin.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
+// Copyright Project Harbor Authors
```
Comment (reasonerjt) [inline src/common/ram/cabin.go]: The file name should be `casbin.go`

File: src/common/ram/cabin.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+# Policy definition
+[policy_definition]
+p = sub, obj, act
+
+# Role definition
+[role_definition]
+g = _, _
+
+# Policy effect
```
Comment (reasonerjt) [inline src/common/ram/cabin.go]: Add more policy effect to support deny, though we don't need it now...

File: src/common/ram/cabin.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+[policy_effect]
+e = some(where (p.eft == allow))
+
+# Matchers
+[matchers]
+m = g(r.sub, p.sub) && r.obj == p.obj && (r.act == p.act || p.act == '*')
```
Comment (reasonerjt) [inline src/common/ram/cabin.go]: Add more variation in obj matching.

File: src/common/ram/cabin.go (hunk 1 above)
Comment (heww) [reply src/common/ram/cabin.go]: file renamed

File: src/common/ram/cabin.go (hunk 2 above)
Comment (heww) [reply src/common/ram/cabin.go]: policy definition add effect

File: src/common/ram/cabin.go (hunk 3 above)
Comment (heww) [reply src/common/ram/cabin.go]: obj matching changed to keyMatch2 which suport * and :var

File: src/common/ram/casbin.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+func (a *userAdapter) LoadPolicy(model model.Model) error {
+	for _, line := range a.getUserAllPolicyLines() {
+		persist.LoadPolicyLine(line, model)
+	}
+
+	return nil
+}
+
+func (a *userAdapter) SavePolicy(model model.Model) error {
+	return nil
```
Comment (steven-zou) [inline src/common/ram/casbin.go]: Suggest return errors.New("Not implemented")

File: src/common/ram/ram.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+// Resource the type of resource
+type Resource string
+
+func (res Resource) String() string {
+	return string(res)
+}
+
+// Subresource returns subresource
+func (res Resource) Subresource(resources ...Resource) Resource {
+	els := []string{res.String()}
```
Comment (steven-zou) [inline src/common/ram/ram.go]: Can we change `els` with another meaningful term?

Comment (heww) [reply src/common/ram/ram.go]: els changed to elements

File: src/common/ram/casbin.go (hunk 4 above)
Comment (heww) [reply src/common/ram/casbin.go]: fixed

**Pull request merged**